File: 01-machine/06-file-descriptors/java/FdLimit.java
// Layer 1 - File descriptors, Java version.
// Same story as the Node/Rust versions: no getrlimit binding in the
// standard library, so this reads /proc/self/limits directly. Note what
// throws here isn't a raw errno -- it's java.io.FileNotFoundException with
// a message mentioning "Too many open files", because FileInputStream
// wraps the open(2) failure in Java's general-purpose IO exception
// hierarchy rather than surfacing EMFILE as a distinct, checkable code the
// way Python's OSError.errno or Go's syscall.EMFILE do. That's worth
// noticing on its own: catching this specific failure in Java means
// string-matching the exception message, which is less robust than the
// other languages' approach.
import java.io.FileInputStream;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;

public class FdLimit {
    static String readSoftLimit() throws IOException {
        return Files.lines(Paths.get("/proc/self/limits"))
                .filter(l -> l.startsWith("Max open files"))
                .findFirst()
                .orElse("(unknown -- not on Linux?)");
    }

    public static void main(String[] args) throws IOException {
        System.out.println(readSoftLimit());

        List<FileInputStream> fds = new ArrayList<>();
        try {
            while (true) {
                fds.add(new FileInputStream("/dev/null"));
            }
        } catch (IOException e) {
            System.out.printf("hit %s after opening %d fds%n", e.getClass().getSimpleName() + ": " + e.getMessage(), fds.size());
        } finally {
            for (FileInputStream f : fds) f.close();
            System.out.printf("closed all %d fds; process is healthy again%n", fds.size());
        }
    }
}

File: 01-machine/06-file-descriptors/nodejs/fd_limit.js
// Layer 1 - Same experiment as the Python version, from Node's side. Node
// has no direct getrlimit() binding in the standard library, so we read
// /proc/self/limits directly (Linux-specific, which is fine for this
// sandbox) to report the ceiling we're about to hit.
const fs = require("fs");

function readSoftLimit() {
  const limits = fs.readFileSync("/proc/self/limits", "utf8");
  const line = limits.split("\n").find((l) => l.startsWith("Max open files"));
  return line ? line.trim() : "(unknown -- not on Linux?)";
}

console.log(readSoftLimit());

const fds = [];
try {
  while (true) {
    fds.push(fs.openSync("/dev/null", "r"));
  }
} catch (e) {
  console.log(`hit ${e.code} ('too many open files') after opening ${fds.length} fds`);
} finally {
  for (const fd of fds) fs.closeSync(fd);
  console.log(`closed all ${fds.length} fds; process is healthy again`);
}
